File: LeetCode/82-remove-duplicates-from-sorted-list-ii/Solution.java
/**
 * Removes all duplicate occurrences from a sorted linked list,
 * ensuring only distinct values remain.
 *
 * This approach uses a sentinel node to simplify edge case handling,
 * and a tracking pointer (`prevUnique`) to manage deletions efficiently.
 *
 * @param head The head of the sorted linked list.
 * @return The modified list with all duplicate elements removed.
 */
class Solution {
  /**
   * Deletes all nodes containing duplicate values from a sorted linked list.
   *
   * @param head The head of the linked list.
   * @return The head of the modified list with only distinct values.
   */
  public ListNode deleteDuplicates(ListNode head) {
    // Edge case: If the list is empty, return null immediately.
    if (head == null) {
      return null;
    }

    // Sentinel node simplifies edge cases where the head may be removed.
    ListNode sentinel = new ListNode(0, head);
    
    // Tracks the last distinct node before duplicate sequences.
    ListNode prevUnique = sentinel; 

    // Traverse the list.
    while (head != null) {
      // If a duplicate sequence is detected:
      if (head.next != null && head.val == head.next.val) {
        // Move to the last occurrence within the duplicate sequence.
        while (head.next != null && head.val == head.next.val) {
          head = head.next;
        }
        // Remove all occurrences of the duplicate value.
        prevUnique.next = head.next;
      } else {
        // Move prevUnique forward only when a unique value is found.
        prevUnique = prevUnique.next;
      }

      // Advance the head pointer to the next node.
      head = head.next;
    }

    // Return the modified list, skipping the sentinel node.
    return sentinel.next;
  }
}

File: LeetCode/82-remove-duplicates-from-sorted-list-ii/solution.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function deleteDuplicates(head: ListNode | null): ListNode | null {
  if (!head) {
    return null;
  }

  // Sentinel node simplifies edge cases
  let sentinel: ListNode = new ListNode(-1, head); // Using -1 to signal a unusuable data
  let prevNode: ListNode = sentinel; // Tracks last unique node before duplicates

  while (head) {
    if (head.next && head.val === head.next.val) {
      while (head.next && head.val === head.next.val) {
        head = head?.next;
      }
      prevNode.next = head.next;
    } else {
      // Advance prevNode only when a unique element is found
      prevNode = prevNode.next!;
    }

    head = head.next;
  }

  return sentinel.next;
};